normalizer_lib.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
+"""Library of language-agnostic FST rewrite rules to normalize text."""
+
+import unicodedata
+from pynini import *
+from config import *
+
+LANGUAGE = af
+
+GRAPHEMES = LANGUAGE.GRAPHEMES
+
+INITIAL_PUNCTUATION = LANGUAGE.INITIAL_PUNCTUATION
+FINAL_PUNCTUATION = LANGUAGE.FINAL_PUNCTUATION
+OTHER_PUNCTUATION = union(r"\[", r"\]")
+PUNCTUATION = union(INITIAL_PUNCTUATION, FINAL_PUNCTUATION, OTHER_PUNCTUATION)
+NUMBERS = union("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")
+SPACE = acceptor(" ")
+
+SIGMA_STAR = union(*("[{}]".format(i) for i in range(1, 256))
+                   ).optimize().closure()
+
+
+# Step 1: Remove all extra whitespace between words
+
+REMOVE_EXTRA_WHITESPACE = transducer(SPACE, "")
+
+DO_REMOVE_EXTRA_WHITESPACE = cdrewrite(
```

Comment:
Instead of naming the cdrewrites "DO_", let's name them what they're doing, like "REMOVE_EXTRA_WHITESPACE", then rename the transducer to something more simple?

Reply:
Makes sense! I think I just carried over the naming from the starter project, where there was a SOFT_MUTATION string_map and a DO_SOFT_MUTATION cdrewrite rule. This way is definitely more coherent though :)